_posts/2025-06-27-Enhancing-Search-Intelligence:-Introduction-to-ML-Inference-Processors-in-OpenSearch.md

```diff
@@ -0,0 +1,536 @@
+
+## What are ML Inference Processors?
+
+OpenSearch provides three types of ML inference processors, each serving a specific purpose in the ingest pipeline or search pipeline:
+
+* ML Inference Ingest Processors: (Released 2.14): These processors run model inference during document indexing, allowing you to enrich or transform documents with ML model predictions before they're stored. For example, you could generate text embeddings or add sentiment labels to documents during indexing.
+* ML Search Request Processors: (Released 2.16): These processors run model inference while operating on search queries before they're executed, enabling query enhancement or transformation. They can rewrite queries based on ML model output, making them particularly useful for query understanding and expansion.
+* ML Search Response Processors(Released 2.16): These processors run model inference and enriched search results before they're returned to the user, enabling features like result reranking or adding ML-generated summaries to search results.
+
+## Key Benefits
+
+1. Flexibility: Support for both local and external ML models
+2. Real-time Processing: ML inference happens during ingest and search operations
+3. Customizable: Extensive configuration options for input/output mapping
+4. Easy-to-apply: Configure once in the search/ingest pipeline, auto applies in every ingest/search operations
+
+## Common Use Cases
+
+ML inference processors enable various advanced search capabilities:
+
+* Semantic Search: Generate and search with text embeddings
+* Multi-Modal Search: Combine text and image features
+* Query Understanding: Enhance queries with ML-based understanding
+* Result Ranking: Improve result relevance using ML models
+* Document Enrichment: Add ML-generated metadata during indexing
+
+## A Simple Example: Search And Summarization Pipeline
+
+Let's walk through a practical example of using ML inference search response processors to implement semantic search with text embedding model and generate summarization with Large Language Model(LLM). 
+
+### Step 1:  Register models in OpenSearch
+
+We would need a text embedding model and a large language model for summarization, in this example, we would use amazon.titan-embed-text-v2:0  and  us.anthropic.claude-3-7-sonnet-20250219-v1:0 . 
+
+Register a text embedding model: 
+
+```json
+POST /_plugins/_ml/connectors/_create
+{
+  "name": "Amazon Bedrock: claude-3-7-sonnet-20250219-v1:0",
```

Comment:
@mingshl This should be Titan embedding, correct? Please verify the name and description of this request.

Reply:
correct, I changed the description in the new commit